File: pc3/monitoreo.py
# monitoreo.py - servicio de monitoreo y consulta - pc3
#
# este es el servicio que usa el usuario (operador) para interactuar
# con el sistema de tráfico. tiene un menú por consola con estas opciones:
#
#   1. ver estado actual de una intersección (pregunta a analítica)
#   2. consulta histórica entre fechas (pregunta a bd principal)
#   3. ver resumen de estados de congestión (pregunta a bd)
#   4. forzar cambio de semáforo - ambulancia (envía a analítica)
#   5. ver throughput del sistema (pregunta a bd)
#   6. consultar datos de intersección en bd
#
# tolerancia a fallos:
# si la bd principal (pc3) no responde, este servicio cambia
# automáticamente para consultar a la bd réplica (pc2).
# el programa no se cierra, sigue funcionando normal.
#
# autores: miguel angel acuña, juan david acuña, y samuel felipe manrique - sistemas distribuidos 2026-10

import zmq
import json
import time
from datetime import datetime, timezone

# ============================================================
# CONFIGURACIÓN DE RED - Cambiar según las IPs
# ============================================================
ANALITICA_IP = "10.43.98.199"      # PC2 - servicio de analítica
BD_PRINCIPAL_IP = "10.43.99.183"   # PC3 - BD principal (esta máquina)
BD_REPLICA_IP = "10.43.98.199"     # PC2 - BD réplica (fallback)

TIMEOUT = 5000  # Timeout en milisegundos (5 segundos)

# TODO: Implementar CurveZMQ para la entrega final

# Esta variable indica si estamos usando la réplica por fallo
usando_replica = False


def crear_socket(contexto, tipo, endpoint):
    # crea un socket zmq simple con timeout configurado.
    socket = contexto.socket(tipo)
    socket.setsockopt(zmq.RCVTIMEO, TIMEOUT)
    socket.setsockopt(zmq.SNDTIMEO, TIMEOUT)
    socket.setsockopt(zmq.LINGER, 0)
    socket.connect(endpoint)
    return socket


def consultar_bd(contexto, socket_bd, consulta, endpoint_actual):
    # envía una consulta a la bd y maneja el failover automático.
    # 
    # si la bd principal no responde, cierro el socket viejo,
    # creo uno nuevo conectado a la bd réplica y reintento.
    # el programa nunca se cierra por esto.
    global usando_replica

    try:
        # Intento enviar y recibir la consulta
        socket_bd.send_json(consulta)
        respuesta = socket_bd.recv_json()
        return respuesta, socket_bd, endpoint_actual

    except (zmq.Again, zmq.ZMQError) as error:
        # La BD no respondió a tiempo
        if not usando_replica:
            # ====================================================
            # ENMASCARAMIENTO DE FALLOS TRANSPARENTE:
            # La BD principal no responde, me reconecto a la réplica.
            # El usuario puede seguir usando el sistema normal.
            # ====================================================
            print(f"\n[MONITOREO] !!! BD Principal no responde: {error}")
            print(f"[MONITOREO] !!! Cambiando a BD Réplica (PC2)...")

            # Cierro el socket viejo y creo uno nuevo a la réplica
            try:
                socket_bd.close()
            except:
                pass

            endpoint_nuevo = f"tcp://{BD_REPLICA_IP}:5564"
            socket_bd = crear_socket(contexto, zmq.REQ, endpoint_nuevo)
            usando_replica = True
            endpoint_actual = endpoint_nuevo

            # Reintento la consulta en la réplica
            try:
                socket_bd.send_json(consulta)
                respuesta = socket_bd.recv_json()
                print(f"[MONITOREO] Consulta exitosa en BD Réplica")
                return respuesta, socket_bd, endpoint_actual
            except (zmq.Again, zmq.ZMQError) as error2:
                print(f"[MONITOREO] BD Réplica tampoco responde: {error2}")
                # Recreo el socket para la próxima vez
                try:
                    socket_bd.close()
                except:
                    pass
                socket_bd = crear_socket(contexto, zmq.REQ, endpoint_nuevo)
                return {"status": "ERROR", "mensaje": "Ninguna BD disponible"}, \
                       socket_bd, endpoint_actual
        else:
            # Ya estamos en la réplica y tampoco responde
            print(f"[MONITOREO] BD Réplica no responde: {error}")
            try:
                socket_bd.close()
            except:
                pass
            socket_bd = crear_socket(contexto, zmq.REQ, f"tcp://{BD_REPLICA_IP}:5564")
            return {"status": "ERROR", "mensaje": "BD Réplica no disponible"}, \
                   socket_bd, endpoint_actual


def mostrar_menu():
    # muestra el menú principal.
    print("\n" + "=" * 55)
    print("   MONITOREO Y CONSULTA - Sistema de Tráfico")
    print("=" * 55)
    print("   1. Estado actual de una intersección")
    print("   2. Consulta histórica (rango de fechas)")
    print("   3. Resumen de estados de congestión")
    print("   4. Forzar semáforo (ambulancia)")
    print("   5. Ver throughput del sistema")
    print("   6. Consultar intersección en BD")
    print("   0. Salir")
    print("=" * 55)
    if usando_replica:
        print("   !! MODO FAILOVER: Usando BD Réplica (PC2)")
    else:
        print("   BD Principal (PC3) conectada")
    print()


def main():
    global usando_replica

    print("=" * 55)
    print("  SERVICIO DE MONITOREO - PC3")
    print("=" * 55)
    print(f"  Analítica:    tcp://{ANALITICA_IP}:5560")
    print(f"  BD Principal: tcp://{BD_PRINCIPAL_IP}:5571")
    print(f"  BD Réplica:   tcp://{BD_REPLICA_IP}:5564 (failover)")
    print("=" * 55)

    contexto = zmq.Context()

    # Socket REQ para hablar con la analítica (comandos directos)
    socket_analitica = crear_socket(contexto, zmq.REQ, f"tcp://{ANALITICA_IP}:5560")
    print(f"[MONITOREO] Conectado a analítica")

    # Socket REQ para consultas a la BD principal
    endpoint_bd = f"tcp://{BD_PRINCIPAL_IP}:5571"
    socket_bd = crear_socket(contexto, zmq.REQ, endpoint_bd)
    print(f"[MONITOREO] Conectado a BD principal")

    print("[MONITOREO] Servicio listo.\n")

    while True:
        try:
            mostrar_menu()
            opcion = input("   Opción: ").strip()

            # ============================================
            # OPCIÓN 1: Estado actual (pregunta a analítica)
            # ============================================
            if opcion == "1":
                inter = input("   Intersección (ej. INT-C5): ").strip().upper()
                if not inter:
                    print("   Intersección inválida")
                    continue

                print(f"\n[MONITOREO] Consultando estado de {inter}...")

                try:
                    socket_analitica.send_json({
                        "tipo": "ESTADO_ACTUAL",
                        "interseccion": inter
                    })
                    resp = socket_analitica.recv_json()

                    if resp.get("status") == "OK" and "Q" in resp:
                        print(f"\n   Intersección: {resp.get('interseccion')}")
                        print(f"   Q (Cola):       {resp.get('Q')}")
                        print(f"   Vp (Velocidad): {resp.get('Vp')} km/h")
                        print(f"   D (Densidad):   {resp.get('D')} veh/km")
                        print(f"   Estado:         {resp.get('estado_trafico')}")
                    else:
                        print(f"   {resp.get('mensaje', 'Sin datos')}")
                except (zmq.Again, zmq.ZMQError) as e:
                    print(f"   Error: analítica no responde ({e})")
                    # Recreo el socket para que no quede en estado raro
                    try:
                        socket_analitica.close()
                    except:
                        pass
                    socket_analitica = crear_socket(contexto, zmq.REQ,
                                                     f"tcp://{ANALITICA_IP}:5560")

            # ============================================
            # OPCIÓN 2: Consulta histórica (pregunta a BD)
            # ============================================
            elif opcion == "2":
                print("   Formato: YYYY-MM-DDTHH:MM:SSZ")
                inicio = input("   Fecha inicio: ").strip()
                fin = input("   Fecha fin:    ").strip()
                if not inicio or not fin:
                    print("   Fechas inválidas")
                    continue

                consulta = {
                    "tipo": "CONSULTA_HISTORICA",
                    "fecha_inicio": inicio,
                    "fecha_fin": fin
                }

                resp, socket_bd, endpoint_bd = consultar_bd(
                    contexto, socket_bd, consulta, endpoint_bd
                )

                print(f"\n   Fuente: {resp.get('fuente', '?')}")
                if resp.get("status") == "OK":
                    total = resp.get("total", 0)
                    print(f"   Registros: {total}")
                    for r in resp.get("resultados", [])[:10]:
                        print(f"   | {r['timestamp']} | {r['interseccion']} | "
                              f"{r['estado_trafico']} | Q={r['Q']} Vp={r['Vp']} D={r['D']}")
                    if total > 10:
                        print(f"   ... ({total} total, mostrando 10)")
                else:
                    print(f"   Error: {resp.get('mensaje')}")

            # ============================================
            # OPCIÓN 3: Resumen de estados (pregunta a BD)
            # ============================================
            elif opcion == "3":
                consulta = {"tipo": "CONSULTA_ESTADOS"}
                resp, socket_bd, endpoint_bd = consultar_bd(
                    contexto, socket_bd, consulta, endpoint_bd
                )

                print(f"\n   Fuente: {resp.get('fuente', '?')}")
                if resp.get("status") == "OK":
                    estados = resp.get("resumen_estados", {})
                    for estado, cantidad in estados.items():
                        barra = "#" * min(cantidad, 50)
                        print(f"   {estado:15s} | {cantidad:5d} | {barra}")
                else:
                    print(f"   Error: {resp.get('mensaje')}")

            # ============================================
            # OPCIÓN 4: Priorización / ambulancia
            # ============================================
            elif opcion == "4":
                inter = input("   Intersección (ej. INT-C5): ").strip().upper()
                motivo = input("   Motivo (ej. Ambulancia): ").strip()
                if not inter:
                    print("   Intersección inválida")
                    continue
                if not motivo:
                    motivo = "Priorización por operador"

                # Mido el tiempo para calcular la latencia
                tiempo_inicio = time.time()

                print(f"\n[MONITOREO] Enviando priorización para {inter}...")

                try:
                    socket_analitica.send_json({
                        "tipo": "PRIORIZACION",
                        "interseccion": inter,
                        "motivo": motivo,
                        "timestamp_solicitud": datetime.now(timezone.utc).strftime(
                            "%Y-%m-%dT%H:%M:%SZ")
                    })
                    resp = socket_analitica.recv_json()

                    # Calculo la latencia (tiempo de respuesta del semáforo)
                    latencia_ms = (time.time() - tiempo_inicio) * 1000

                    if resp.get("status") == "OK":
                        print(f"   {resp.get('mensaje')}")
                        cmd = resp.get("comando", {})
                        print(f"   Acción: {cmd.get('accion')}")
                        print(f"   Verde: {cmd.get('duracion_verde')}s")
                        print(f"   Latencia: {latencia_ms:.1f} ms")
                    else:
                        print(f"   Error: {resp.get('mensaje')}")
                except (zmq.Again, zmq.ZMQError) as e:
                    print(f"   Error: analítica no responde ({e})")
                    try:
                        socket_analitica.close()
                    except:
                        pass
                    socket_analitica = crear_socket(contexto, zmq.REQ,
                                                     f"tcp://{ANALITICA_IP}:5560")

            # ============================================
            # OPCIÓN 5: Throughput (pregunta a BD)
            # ============================================
            elif opcion == "5":
                consulta = {"tipo": "CONSULTA_THROUGHPUT"}
                resp, socket_bd, endpoint_bd = consultar_bd(
                    contexto, socket_bd, consulta, endpoint_bd
                )

                print(f"\n   Fuente: {resp.get('fuente', '?')}")
                if resp.get("status") == "OK":
                    print(f"   Total registros en BD: {resp.get('total_registros')}")
                else:
                    print(f"   Error: {resp.get('mensaje')}")

            # ============================================
            # OPCIÓN 6: Consultar intersección en BD
            # ============================================
            elif opcion == "6":
                inter = input("   Intersección (ej. INT-C5): ").strip().upper()
                if not inter:
                    print("   Intersección inválida")
                    continue

                consulta = {
                    "tipo": "CONSULTA_INTERSECCION",
                    "interseccion": inter
                }
                resp, socket_bd, endpoint_bd = consultar_bd(
                    contexto, socket_bd, consulta, endpoint_bd
                )

                print(f"\n   Fuente: {resp.get('fuente', '?')}")
                if resp.get("status") == "OK":
                    total = resp.get("total", 0)
                    print(f"   Registros: {total}")
                    for r in resp.get("resultados", []):
                        print(f"   | {r['timestamp']} | {r['tipo_sensor']} | "
                              f"{r['estado_trafico']} | Q={r['Q']} Vp={r['Vp']} D={r['D']}")
                else:
                    print(f"   Error: {resp.get('mensaje')}")

            # ============================================
            # OPCIÓN 0: Salir
            # ============================================
            elif opcion == "0":
                print("\n[MONITOREO] Cerrando...")
                break
            else:
                print("   Opción inválida")

        except KeyboardInterrupt:
            print("\n\n[MONITOREO] Cerrando...")
            break
        except EOFError:
            break

    # Cierro todo
    socket_analitica.close()
    socket_bd.close()
    contexto.term()
    print("[MONITOREO] Listo.")


if __name__ == "__main__":
    main()

File: java/pc3/src/main/java/Monitoreo.java
// monitoreo.java - servicio de monitoreo y consulta - pc3
//
// este es el servicio que usa el usuario (operador) para interactuar
// con el sistema de trafico. tiene un menu por consola con estas opciones:
//
//   1. ver estado actual de una interseccion (pregunta a analitica)
//   2. consulta historica entre fechas (pregunta a bd principal)
//   3. ver resumen de estados de congestion (pregunta a bd)
//   4. forzar cambio de semaforo - ambulancia (envia a analitica)
//   5. ver throughput del sistema (pregunta a bd)
//   6. consultar datos de interseccion en bd
//
// tolerancia a fallos:
// si la bd principal (pc3) no responde, este servicio cambia
// automaticamente para consultar a la bd replica (pc2).
// el programa no se cierra, sigue funcionando normal.
//
// autores: miguel angel acuna, juan david acuna, y samuel felipe manrique - sistemas distribuidos 2026-10

import org.zeromq.SocketType;
import org.zeromq.ZContext;
import org.zeromq.ZMQ;
import org.json.JSONObject;
import org.json.JSONArray;

import java.time.Instant;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.Scanner;

public class Monitoreo {

    // ============================================================
    // configuracion de red - cambiar segun las ips
    // ============================================================
    static String ANALITICA_IP = "10.43.98.199";      // pc2 - servicio de analitica
    static String BD_PRINCIPAL_IP = "10.43.99.183";   // pc3 - bd principal (esta maquina)
    static String BD_REPLICA_IP = "10.43.98.199";     // pc2 - bd replica (fallback)

    static int TIMEOUT = 5000;  // timeout en milisegundos (5 segundos)

    // esta variable indica si estamos usando la replica por fallo
    static boolean usandoReplica = false;

    // sockets y contexto como variables estaticas para poder modificarlos en funciones
    static ZContext contexto;
    static ZMQ.Socket socketAnalitica;
    static ZMQ.Socket socketBd;
    static String endpointBd;

    // crea un socket zmq simple con timeout configurado.
    static ZMQ.Socket crearSocket(SocketType tipo, String endpoint) {
        ZMQ.Socket socket = contexto.createSocket(tipo);
        socket.setReceiveTimeOut(TIMEOUT);
        socket.setSendTimeOut(TIMEOUT);
        socket.setLinger(0);
        socket.connect(endpoint);
        return socket;
    }

    // envia una consulta a la bd y maneja el failover automatico.
    //
    // si la bd principal no responde, cierro el socket viejo,
    // creo uno nuevo conectado a la bd replica y reintento.
    // el programa nunca se cierra por esto.
    static JSONObject consultarBd(JSONObject consulta) {
        try {
            // intento enviar y recibir la consulta
            socketBd.send(consulta.toString());
            String respStr = socketBd.recvStr();
            if (respStr != null) {
                return new JSONObject(respStr);
            }
            throw new Exception("timeout");
        } catch (Exception error) {
            // la bd no respondio a tiempo
            if (!usandoReplica) {
                // ====================================================
                // enmascaramiento de fallos transparente:
                // la bd principal no responde, me reconecto a la replica.
                // el usuario puede seguir usando el sistema normal.
                // ====================================================
                System.out.println("\n[MONITOREO] !!! BD Principal no responde: " + error.getMessage());
                System.out.println("[MONITOREO] !!! Cambiando a BD Replica (PC2)...");

                // cierro el socket viejo y creo uno nuevo a la replica
                try { socketBd.close(); } catch (Exception e) { /* nada */ }

                endpointBd = "tcp://" + BD_REPLICA_IP + ":5564";
                socketBd = crearSocket(SocketType.REQ, endpointBd);
                usandoReplica = true;

                // reintento la consulta en la replica
                try {
                    socketBd.send(consulta.toString());
                    String respStr = socketBd.recvStr();
                    if (respStr != null) {
                        System.out.println("[MONITOREO] Consulta exitosa en BD Replica");
                        return new JSONObject(respStr);
                    }
                    throw new Exception("timeout replica");
                } catch (Exception error2) {
                    System.out.println("[MONITOREO] BD Replica tampoco responde: " + error2.getMessage());
                    // recreo el socket para la proxima vez
                    try { socketBd.close(); } catch (Exception e) { /* nada */ }
                    socketBd = crearSocket(SocketType.REQ, endpointBd);
                    JSONObject err = new JSONObject();
                    err.put("status", "ERROR");
                    err.put("mensaje", "Ninguna BD disponible");
                    return err;
                }
            } else {
                // ya estamos en la replica y tampoco responde
                System.out.println("[MONITOREO] BD Replica no responde: " + error.getMessage());
                try { socketBd.close(); } catch (Exception e) { /* nada */ }
                socketBd = crearSocket(SocketType.REQ, "tcp://" + BD_REPLICA_IP + ":5564");
                JSONObject err = new JSONObject();
                err.put("status", "ERROR");
                err.put("mensaje", "BD Replica no disponible");
                return err;
            }
        }
    }

    // muestra el menu principal.
    static void mostrarMenu() {
        System.out.println("\n=======================================================");
        System.out.println("   MONITOREO Y CONSULTA - Sistema de Trafico");
        System.out.println("=======================================================");
        System.out.println("   1. Estado actual de una interseccion");
        System.out.println("   2. Consulta historica (rango de fechas)");
        System.out.println("   3. Resumen de estados de congestion");
        System.out.println("   4. Forzar semaforo (ambulancia)");
        System.out.println("   5. Ver throughput del sistema");
        System.out.println("   6. Consultar interseccion en BD");
        System.out.println("   0. Salir");
        System.out.println("=======================================================");
        if (usandoReplica) {
            System.out.println("   !! MODO FAILOVER: Usando BD Replica (PC2)");
        } else {
            System.out.println("   BD Principal (PC3) conectada");
        }
        System.out.println();
    }

    public static void main(String[] args) {
        System.out.println("=======================================================");
        System.out.println("  SERVICIO DE MONITOREO - PC3");
        System.out.println("=======================================================");
        System.out.println("  Analitica:    tcp://" + ANALITICA_IP + ":5560");
        System.out.println("  BD Principal: tcp://" + BD_PRINCIPAL_IP + ":5571");
        System.out.println("  BD Replica:   tcp://" + BD_REPLICA_IP + ":5564 (failover)");
        System.out.println("=======================================================");

        contexto = new ZContext();

        // socket req para hablar con la analitica (comandos directos)
        socketAnalitica = crearSocket(SocketType.REQ, "tcp://" + ANALITICA_IP + ":5560");
        System.out.println("[MONITOREO] Conectado a analitica");

        // socket req para consultas a la bd principal
        endpointBd = "tcp://" + BD_PRINCIPAL_IP + ":5571";
        socketBd = crearSocket(SocketType.REQ, endpointBd);
        System.out.println("[MONITOREO] Conectado a BD principal");

        System.out.println("[MONITOREO] Servicio listo.\n");

        Scanner scanner = new Scanner(System.in);

        while (true) {
            try {
                mostrarMenu();
                System.out.print("   Opcion: ");
                String opcion = scanner.nextLine().trim();

                // ============================================
                // opcion 1: estado actual (pregunta a analitica)
                // ============================================
                if (opcion.equals("1")) {
                    System.out.print("   Interseccion (ej. INT-C5): ");
                    String inter = scanner.nextLine().trim().toUpperCase();
                    if (inter.isEmpty()) { System.out.println("   Interseccion invalida"); continue; }

                    System.out.println("\n[MONITOREO] Consultando estado de " + inter + "...");
                    try {
                        JSONObject req = new JSONObject();
                        req.put("tipo", "ESTADO_ACTUAL");
                        req.put("interseccion", inter);
                        socketAnalitica.send(req.toString());
                        String respStr = socketAnalitica.recvStr();
                        if (respStr == null) throw new Exception("timeout");
                        JSONObject resp = new JSONObject(respStr);

                        if (resp.optString("status").equals("OK") && resp.has("Q")) {
                            System.out.println("\n   Interseccion: " + resp.optString("interseccion"));
                            System.out.println("   Q (Cola):       " + resp.opt("Q"));
                            System.out.println("   Vp (Velocidad): " + resp.opt("Vp") + " km/h");
                            System.out.println("   D (Densidad):   " + resp.opt("D") + " veh/km");
                            System.out.println("   Estado:         " + resp.optString("estado_trafico"));
                        } else {
                            System.out.println("   " + resp.optString("mensaje", "Sin datos"));
                        }
                    } catch (Exception e) {
                        System.out.println("   Error: analitica no responde (" + e.getMessage() + ")");
                        try { socketAnalitica.close(); } catch (Exception ex) { /* nada */ }
                        socketAnalitica = crearSocket(SocketType.REQ, "tcp://" + ANALITICA_IP + ":5560");
                    }

                // ============================================
                // opcion 2: consulta historica (pregunta a bd)
                // ============================================
                } else if (opcion.equals("2")) {
                    System.out.println("   Formato: YYYY-MM-DDTHH:MM:SSZ");
                    System.out.print("   Fecha inicio: ");
                    String inicio = scanner.nextLine().trim();
                    System.out.print("   Fecha fin:    ");
                    String fin = scanner.nextLine().trim();
                    if (inicio.isEmpty() || fin.isEmpty()) { System.out.println("   Fechas invalidas"); continue; }

                    JSONObject consulta = new JSONObject();
                    consulta.put("tipo", "CONSULTA_HISTORICA");
                    consulta.put("fecha_inicio", inicio);
                    consulta.put("fecha_fin", fin);

                    JSONObject resp = consultarBd(consulta);
                    System.out.println("\n   Fuente: " + resp.optString("fuente", "?"));
                    if (resp.optString("status").equals("OK")) {
                        int total = resp.optInt("total", 0);
                        System.out.println("   Registros: " + total);
                        JSONArray resultados = resp.optJSONArray("resultados");
                        if (resultados != null) {
                            int limit = Math.min(10, resultados.length());
                            for (int i = 0; i < limit; i++) {
                                JSONObject r = resultados.getJSONObject(i);
                                System.out.println("   | " + r.optString("timestamp") + " | " + r.optString("interseccion")
                                        + " | " + r.optString("estado_trafico") + " | Q=" + r.opt("Q")
                                        + " Vp=" + r.opt("Vp") + " D=" + r.opt("D"));
                            }
                            if (total > 10) System.out.println("   ... (" + total + " total, mostrando 10)");
                        }
                    } else {
                        System.out.println("   Error: " + resp.optString("mensaje"));
                    }

                // ============================================
                // opcion 3: resumen de estados (pregunta a bd)
                // ============================================
                } else if (opcion.equals("3")) {
                    JSONObject consulta = new JSONObject();
                    consulta.put("tipo", "CONSULTA_ESTADOS");
                    JSONObject resp = consultarBd(consulta);
                    System.out.println("\n   Fuente: " + resp.optString("fuente", "?"));
                    if (resp.optString("status").equals("OK")) {
                        JSONObject estados = resp.optJSONObject("resumen_estados");
                        if (estados != null) {
                            for (String key : estados.keySet()) {
                                int cantidad = estados.getInt(key);
                                String barra = "#".repeat(Math.min(cantidad, 50));
                                System.out.printf("   %-15s | %5d | %s%n", key, cantidad, barra);
                            }
                        }
                    } else {
                        System.out.println("   Error: " + resp.optString("mensaje"));
                    }

                // ============================================
                // opcion 4: priorizacion / ambulancia
                // ============================================
                } else if (opcion.equals("4")) {
                    System.out.print("   Interseccion (ej. INT-C5): ");
                    String inter = scanner.nextLine().trim().toUpperCase();
                    System.out.print("   Motivo (ej. Ambulancia): ");
                    String motivo = scanner.nextLine().trim();
                    if (inter.isEmpty()) { System.out.println("   Interseccion invalida"); continue; }
                    if (motivo.isEmpty()) motivo = "Priorizacion por operador";

                    // mido el tiempo para calcular la latencia
                    long tiempoInicio = System.currentTimeMillis();

                    System.out.println("\n[MONITOREO] Enviando priorizacion para " + inter + "...");
                    try {
                        JSONObject req = new JSONObject();
                        req.put("tipo", "PRIORIZACION");
                        req.put("interseccion", inter);
                        req.put("motivo", motivo);
                        req.put("timestamp_solicitud", timestampAhora());
                        socketAnalitica.send(req.toString());
                        String respStr = socketAnalitica.recvStr();
                        if (respStr == null) throw new Exception("timeout");
                        JSONObject resp = new JSONObject(respStr);

                        // calculo la latencia (tiempo de respuesta del semaforo)
                        double latenciaMs = System.currentTimeMillis() - tiempoInicio;

                        if (resp.optString("status").equals("OK")) {
                            System.out.println("   " + resp.optString("mensaje"));
                            JSONObject cmd = resp.optJSONObject("comando");
                            if (cmd != null) {
                                System.out.println("   Accion: " + cmd.optString("accion"));
                                System.out.println("   Verde: " + cmd.optInt("duracion_verde") + "s");
                            }
                            System.out.printf("   Latencia: %.1f ms%n", latenciaMs);
                        } else {
                            System.out.println("   Error: " + resp.optString("mensaje"));
                        }
                    } catch (Exception e) {
                        System.out.println("   Error: analitica no responde (" + e.getMessage() + ")");
                        try { socketAnalitica.close(); } catch (Exception ex) { /* nada */ }
                        socketAnalitica = crearSocket(SocketType.REQ, "tcp://" + ANALITICA_IP + ":5560");
                    }

                // ============================================
                // opcion 5: throughput (pregunta a bd)
                // ============================================
                } else if (opcion.equals("5")) {
                    JSONObject consulta = new JSONObject();
                    consulta.put("tipo", "CONSULTA_THROUGHPUT");
                    JSONObject resp = consultarBd(consulta);
                    System.out.println("\n   Fuente: " + resp.optString("fuente", "?"));
                    if (resp.optString("status").equals("OK")) {
                        System.out.println("   Total registros en BD: " + resp.optInt("total_registros"));
                    } else {
                        System.out.println("   Error: " + resp.optString("mensaje"));
                    }

                // ============================================
                // opcion 6: consultar interseccion en bd
                // ============================================
                } else if (opcion.equals("6")) {
                    System.out.print("   Interseccion (ej. INT-C5): ");
                    String inter = scanner.nextLine().trim().toUpperCase();
                    if (inter.isEmpty()) { System.out.println("   Interseccion invalida"); continue; }

                    JSONObject consulta = new JSONObject();
                    consulta.put("tipo", "CONSULTA_INTERSECCION");
                    consulta.put("interseccion", inter);
                    JSONObject resp = consultarBd(consulta);
                    System.out.println("\n   Fuente: " + resp.optString("fuente", "?"));
                    if (resp.optString("status").equals("OK")) {
                        int total = resp.optInt("total", 0);
                        System.out.println("   Registros: " + total);
                        JSONArray resultados = resp.optJSONArray("resultados");
                        if (resultados != null) {
                            for (int i = 0; i < resultados.length(); i++) {
                                JSONObject r = resultados.getJSONObject(i);
                                System.out.println("   | " + r.optString("timestamp") + " | " + r.optString("tipo_sensor")
                                        + " | " + r.optString("estado_trafico") + " | Q=" + r.opt("Q")
                                        + " Vp=" + r.opt("Vp") + " D=" + r.opt("D"));
                            }
                        }
                    } else {
                        System.out.println("   Error: " + resp.optString("mensaje"));
                    }

                // ============================================
                // opcion 0: salir
                // ============================================
                } else if (opcion.equals("0")) {
                    System.out.println("\n[MONITOREO] Cerrando...");
                    break;
                } else {
                    System.out.println("   Opcion invalida");
                }

            } catch (Exception e) {
                System.out.println("\n\n[MONITOREO] Cerrando...");
                break;
            }
        }

        // cierro todo
        socketAnalitica.close();
        socketBd.close();
        contexto.close();
        System.out.println("[MONITOREO] Listo.");
    }

    static String timestampAhora() {
        return DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'")
                .withZone(ZoneOffset.UTC)
                .format(Instant.now());
    }
}
